README.rst
=======
fzf-wal
=======

.. image:: screenshot.gif

.. image:: https://img.shields.io/pypi/v/fzf_wal.svg
        :target: https://pypi.python.org/pypi/fzf_wal

.. image:: https://img.shields.io/travis/shanedabes/fzf_wal.svg
        :target: https://travis-ci.org/shanedabes/fzf_wal

.. image:: https://readthedocs.org/projects/fzf-wal/badge/?version=latest
        :target: https://fzf-wal.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status


.. image:: https://pyup.io/repos/github/shanedabes/fzf_wal/shield.svg
     :target: https://pyup.io/repos/github/shanedabes/fzf_wal/
     :alt: Updates



pywal theme selector using fzf


* Free software: Apache Software License 2.0
* Documentation: https://fzf-wal.readthedocs.io.


Features
--------

* Preview and select wal themes using fzf menu

Installation
____________

fzf-wal can be installed using:

* `pip`_ (`pip install fzf-wal`)
* From the `AUR`_

.. _pip: https://pypi.org/project/fzf-wal/
.. _AUR: https://aur.archlinux.org/packages/python-fzf-wal/

Credits
-------

This package was created with Cookiecutter_ and the `audreyr/cookiecutter-pypackage`_ project template.

.. _Cookiecutter: https://github.com/audreyr/cookiecutter
.. _`audreyr/cookiecutter-pypackage`: https://github.com/audreyr/cookiecutter-pypackage


HISTORY.rst
=======
History
=======

0.1.2 (2019-05-22)
------------------

* Fix entrypoint loading main function twice

0.1.0 (2019-05-18)
------------------

* First release on PyPI.


tests/test_fzf_wal.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `fzf_wal` package."""

import pytest
import unittest.mock

import os


from fzf_wal import fzf_wal


def open_relative(fn: str) -> str:
    dir_name = os.path.dirname(os.path.realpath(__file__))
    with open(os.path.join(dir_name, fn)) as f:
        return f.read()


@pytest.fixture
def theme_dict():
    return {
        't1': {
            'colors': {'color0': '000000'},
            'special': {'foreground': '000000', 'background': 'ffffff'}
        },
        't2': {
            'colors': {'color0': '000000'},
            'special': {'foreground': '000000', 'background': 'ffffff'}
        }
    }


hex_to_rgb_params = [
    ('#aabbcc', [170, 187, 204]),
    ('#ddeeff', [221, 238, 255]),
    ('112233',  [17,  34,  51])
]


@pytest.mark.parametrize('hexstring,expected', hex_to_rgb_params)
def test_hex_to_rgb(hexstring, expected):
    rgb = fzf_wal.hex_to_rgb(hexstring)

    assert expected == rgb


def test_hex_to_rgb_too_short():
    with pytest.raises(ValueError) as excinfo:
        fzf_wal.hex_to_rgb('aabb')
    assert str(excinfo.value) == 'RGB hex value must be at least 6 chars'


def test_rgb_string():
    expected = r'[38;2;12;34;56mtest'
    out = fzf_wal.rgb_string('test', [12, 34, 56])

    assert expected == out


def test_rgb_bg_string():
    expected = r'[48;2;12;34;56mtest'
    out = fzf_wal.rgb_bg_string('test', [12, 34, 56])

    assert expected == out


def test_escape_colour():
    expected = 'hello[m'
    escaped = fzf_wal.escape_colour('hello')

    assert escaped == expected


def test_color_band():
    expected = open_relative('colour_band.txt')
    colours = [[0, 0, 0]]*18
    band = fzf_wal.colour_band(colours)

    assert expected == band


def test_theme_dict_colours():
    expected = ['a', 'b', 'c']
    in_dict = {'colors': {'color0': 'a', 'color2': 'c', 'color1': 'b'}}

    colours = fzf_wal.theme_dict_colours(in_dict)

    assert colours == expected


def test_name_from_path():
    path = '/a/b/c/dark/theme'
    expected = 'dark/theme'

    name = fzf_wal.name_from_path(path)

    assert name == expected


def test_theme_name_iter(theme_dict):
    expected = open_relative('theme_name.txt')

    assert '\n'.join((fzf_wal.theme_name_iter(theme_dict))) == expected


def test_name_from_selection():
    expected = 'test'
    name = fzf_wal.name_from_selection(' a b c d test ')

    assert name == expected


@unittest.mock.patch('fzf_wal.fzf_wal.iterfzf')
def test_theme_selector(mock_fzf, theme_dict):
    mock_fzf.side_effect = lambda x: list(x)[0]

    assert fzf_wal.theme_selector(theme_dict) == 't1'


@unittest.mock.patch('fzf_wal.fzf_wal.iterfzf')
def test_theme_selector_cancelled(mock_fzf, theme_dict):
    mock_fzf.return_value = None

    assert fzf_wal.theme_selector(theme_dict) is None


@unittest.mock.patch('fzf_wal.fzf_wal.pywal')
@unittest.mock.patch('fzf_wal.fzf_wal.iterfzf')
def test_fzf_cancelled_sysexit(mock_fzf, mock_pywal):
    mock_fzf.return_value = None

    with pytest.raises(SystemExit) as e:
        fzf_wal.main()
    assert e.type == SystemExit
    assert e.value.code == 1


fzf_wal/fzf_wal.py
# -*- coding: utf-8 -*-

"""Main module."""

import os
import sys

import pywal
from iterfzf import iterfzf


def hex_to_rgb(hexstring: str) -> [int, int, int]:
    """ Convert base 16 hex (ffffff) string to rgb decimals [255,255,255] """
    if len(hexstring) < 6:
        raise ValueError('RGB hex value must be at least 6 chars')

    hexchars = hexstring[-6:]
    colours = [hexchars[i:i+2] for i in range(0, 6, 2)]
    rgb = [int(i, 16) for i in colours]

    return rgb


def rgb_string(s: str, rgb: [int, int, int], attr: str = 38) -> str:
    """ Take a rgb decimal list and string, return a shell colour sequence """
    r, g, b = rgb
    out = rf'[{attr};2;{r};{g};{b}m{s}'

    return out


def rgb_bg_string(s: str, rgb: [int, int, int]) -> str:
    """ Return shell colour sequence for background colours """
    return rgb_string(s, rgb, 48)


def escape_colour(s):
    return f'{s}[m'


def colour_band(colours: list) -> str:
    """ Return a string with coloured blocks for all given rgb lists """
    out = ' '.join(rgb_string('▄', rgb) for rgb in colours[:16])

    return out


def theme_dict_colours(d: dict) -> list:
    """ Extract colours from wal theme dict """
    return [d['colors'][k] for k in sorted(d['colors'])]


def name_from_path(path: str) -> str:
    """ extract shade (light/dark) and theme name from path """
    dirname, fn = os.path.split(path)
    _, shade = os.path.split(dirname)
    name, _ = os.path.splitext(fn)

    return f'{shade}/{name}'


def theme_name_iter(theme_dicts: dict):
    """ an iterable containing previews and names of themes  for fzf """
    for name, theme in theme_dicts.items():
        rgbs = [hex_to_rgb(i) for i in theme_dict_colours(theme)]
        fg, bg = theme['special']['foreground'], theme['special']['background']
        fg_rgb, bg_rgb = [hex_to_rgb(i) for i in [fg, bg]]

        band = colour_band(rgbs)
        fg_block = rgb_string("▄", fg_rgb)
        preview = escape_colour(rgb_bg_string(f' {band} {fg_block} ', bg_rgb))

        yield f'{preview} {name}'


def name_from_selection(s: str) -> str:
    """ extract theme name from fzf selection """
    return s.split()[-1].strip()


def theme_selector(theme_dicts: dict) -> str:
    """ Use fzf to select a theme """
    os.environ['FZF_DEFAULT_OPTS'] = '--ansi'
    selected = iterfzf(theme_name_iter(theme_dicts))
    if selected is None:
        return None

    return name_from_selection(selected)


def main():
    # load all pywal themes
    themes = (pywal.theme.list_themes() +
              pywal.theme.list_themes(dark=False) +
              pywal.theme.list_themes_user())
    # create a dictionary of paths
    theme_files = {name_from_path(i): i.path for i in themes}
    # create a dictionary of theme dictionaries
    theme_dicts = {k: pywal.colors.file(v) for k, v in theme_files.items()}

    # select a theme using fzf
    selected = theme_selector(theme_dicts)
    # if no theme was selected, exit with return status 1
    if selected is None:
        sys.exit(1)

    # if theme was selected, load from dict
    theme = theme_dicts[selected]

    # apply theme
    pywal.sequences.send(theme)
    pywal.export.every(theme)
    pywal.reload.env()
